src/passes/AlignmentLowering.cpp

```diff
@@ -222,7 +214,145 @@ struct AlignmentLowering : public WalkerPass<PostWalker<AlignmentLowering>> {
       WASM_UNREACHABLE("invalid size");
     }
     block->finalize();
-    replaceCurrent(block);
+    return block;
+  }
+
+  void visitLoad(Load* curr) {
+    // If unreachable, just remove the load, which removes the unaligned
+    // operation in a trivial way.
+    if (curr->type == Type::unreachable) {
+      replaceCurrent(curr->ptr);
+      return;
+    }
+    if (curr->align == 0 || curr->align == curr->bytes) {
+      // Nothing to do.
+      return;
+    }
+    Builder builder(*getModule());
+    auto type = curr->type.getSingle();
+    Expression* replacement;
+    switch (type) {
+      default:
+        WASM_UNREACHABLE("unhandled unaligned load");
+      case Type::i32:
+        replacement = lowerLoadI32(curr);
+        break;
+      case Type::f32:
+        curr->type = Type::i32;
+        replacement = builder.makeUnary(ReinterpretInt32, lowerLoadI32(curr));
+        break;
+      case Type::i64:
+      case Type::f64:
+        if (type == Type::i64 && curr->bytes != 8) {
+          // A load of <64 bits.
+          curr->type = Type::i32;
+          replacement = builder.makeUnary(
+            curr->signed_ ? ExtendSInt32 : ExtendUInt32, lowerLoadI32(curr));
+          break;
+        }
+        // Load two 32-bit pieces, and combine them.
+        auto temp = builder.addVar(getFunction(), Type::i32);
+        auto* set = builder.makeLocalSet(temp, curr->ptr);
+        Expression* low =
+          lowerLoadI32(builder.makeLoad(4,
+                                        false,
+                                        curr->offset,
+                                        curr->align,
+                                        builder.makeLocalGet(temp, Type::i32),
+                                        Type::i32));
+        low = builder.makeUnary(ExtendUInt32, low);
+        Expression* high =
+          lowerLoadI32(builder.makeLoad(4,
+                                        false,
+                                        curr->offset + 4,
+                                        curr->align,
```

Comment:
I'm not sure the alignment here is correct, e.g. if `curr->align == 8` then only the low half will actually be aligned to 8 bytes.

Reply:
True for `align == 8`, but we only get here if the alignment is less than 8, which leaves 4, 2, or 1, all of which stay the same when adding 4. But good point that this isn't clear, I'll add a comment!